Step-68 : Particle tracking of a massless tracer in a vortical flow

blaisb opened the pull request:
Preliminary version of a step on particle tracking in an analytically defined velocity field.
It is very simple, but it showcases particles in parallel. It uses an analytically defined velocity field to move the particles either by interpolating from an MPI::Vector (e.g. from Trilinos) or analytically from a function. It uses the DiscreteTime class for time-stepping, a simple Euler scheme and ParameterAccessor for parsing (similar to step-70). The only thing missing that would be very interesting would be to add load balancing using something like 1000+1000*nb_particles_per_cell as a weight.  I tried looking over the CellWeight class but I think I will need some help figuring out how to use it outside of it's original intended scope. I think Rene already did something similar for ASPECT.  Furthermore, once we figure out a generic data out for particle properties, we could attach the mpi rank to the particles.

This is a work in progress and there is no documentation yet.

@peterrum @gassmoeller

peterrum commented on examples/step-x/step-x.cc:
```diff
@@ -0,0 +1,542 @@
+        const auto &dh_cell =
+          typename DoFHandler<dim>::cell_iterator(*cell, &fluid_dh);
+        dh_cell->get_dof_indices(dof_indices);
+        const auto pic = particle_handler.particles_in_cell(cell);
+
+        for (unsigned int i = 0; particle != pic.end(); ++particle, ++i)
+          {
```
Couldn't you replace this by an iterator?

peterrum commented on examples/step-x/step-x.cc:
```diff
@@ -0,0 +1,542 @@
+
+                particle_velocity[comp_j.first] +=
+                  fluid_fe.shape_value(j, reference_location) *
+                  field_relevant(dof_indices[j]);
```
I would read the values first from the global vector into a cell-local vector and only operate with `j` here.

peterrum commented:
@gassmoeller 

Two questions regarding the usability of the `ParticleHandler`. Wouldn't it be possible to teach the `get_surrounding_cell()` function to take also a `DoFHandler` and return in this case the right type of iterator? This would simplify the following code
```cpp
        const auto &cell =
          particle->get_surrounding_cell(background_triangulation);
        const auto &dh_cell =
          typename DoFHandler<dim>::cell_iterator(*cell, &fluid_dh);
        dh_cell->get_dof_indices(dof_indices);
```
to:
```cpp
        const auto cell = particle->get_surrounding_cell(fluid_dh);
```
The following code loop over all cells that contain cells and within that cell it loops over all particles:
```cpp
    auto particle = particle_handler.begin();
    while (particle != particle_handler.end())
      {
        const auto &cell = particle->get_surrounding_cell(triangulation);
        const auto pic = particle_handler.particles_in_cell(cell);

        for (unsigned int i = 0; particle != pic.end(); ++particle, ++i) 
          /*do something*/
      }
```
Wouldn't be possible to introduce appropriate iterators? step-70 already contained these lines of code?

If you want, I could help and open PRs regarding to these points.

blaisb commented:
> Looks good!
> 
> Are you planing to also add also an advection solver? I would be fine, if you wouldn't, and you would concentrate on particle-related aspect.

I think I prefer to concentrate on the particle aspects. I think there are already many tutorials that describe solving equations in parallel. I just envision this step has a very simple, barebone step that focuses on everything related to basic operations on particles.

gassmoeller commented:
I am looking into adding load balancing right now. Once I have something working lets have a chat so I can walk you through the steps. Also I will look into your questions Peter.

gassmoeller commented:
Load balancing is in lethe-cfd/dealii#1

@peterrum:
Yes adding another function `get_surrounding_cell` for DoFHandler iterators would be useful and easy to implement. I will try to get to it within the next days.  I opened #10314 to remind us of the issue.
```
    typename Triangulation<dim, spacedim>::cell_iterator
    get_surrounding_cell(
      const Triangulation<dim, spacedim> &triangulation) const;

    typename DoFHandler<dim, spacedim>::cell_iterator
    get_surrounding_cell(
      const DoFHandler<dim, spacedim> &dof_handler) const;
```

The second case is a bit more complicated. I think Bruno changed it already for this step, which is possible here. But the general issue (if I remember correctly) is that the following three ways to do the same thing (iterating over all particles) have very different performance characteristics:
```
for (particle: particles)
{
cell = particle->get_surrounding_cell()
... do something
}

for (cell: cells)
{
particles = particle_handler.find_particles_in_cell(cell)

for (particle: particles)
{
... do something
}
}

for (particle: particles)
{
cell = particle->get_surrounding_cell();
particles_in_cell = particle_handler.find_particles_in_cell(cell)

for (particle_in_cell: particles_in_cell)
{
... do something and advance particle_in_cell **and** particle
}
}
```
I think there were cases where the third version is necessary. Were you suggesting to add new particle iterators that would only iterate over the particles of a single cell? That could cover the third case. I would be interested in hearing how you would do that, my knowledge about iterators is still a bit limited.

blaisb commented:
@peterrum  I reserved step-68 like you suggested by putting it in the PR's name. Is that ok?
I should finish first run of documentation and results tomorrow since Rene added load balancing.
:)!

blaisb commented:
Dear  @gassmoeller and @peterrum. I finished a first full version of the tutorial with documentations. Rene there are two very minor places where I would appreciate your input for the documentation.

The results are on the following youtube links:
https://youtu.be/EbgS5Ch35Xs
https://youtu.be/ubUcsR4ECj4


As always, I am grateful for all your comments and ideas :)

peterrum commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,120 @@
+Bruno Blais (Polytechnique Montréal),
+Toni El Geitani Nehme (Polytechnique Montreal),
+Rene Gassmöller (University of California Davis),
+and Peter Munch
```
You wouldn't have to list me here. You did the work! But if you want to mention me:

```suggestion
and Peter Munch (Technical University of Munich and Helmholtz-Zentrum Geesthacht)
```

peterrum commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,120 @@
+ or they can even be used to simulate the motion of granular matter, as is
+ the case with the Discrete Element Method (DEM) (Blais et al. 2019). In the case
+ of DEM, the resulting model is not related to the finite element method anymore,
+ but just leads to a serie of ordinary differential equation which describes
```
```suggestion
 but just leads to a system of ordinary differential equation which describes
```

peterrum commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,120 @@
+\frac{d \textbf{x}_i}{dt} =\textbf{u}(\textbf{x}_i)
+@f]
+
+where $\textbf{x}_i$ is the postion of particle $i$. In the present step,
```
... and u its verlocity.

peterrum commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,120 @@
+of particle $i$ at time $t+\Delta t$ and $t$, respectively and where $\Delta t$
+is the time step. In the present step, the velocity at the location of particles
+is obtainedin two different fashions:
+- By evaluating the velocity function at the location of the particles
```
```suggestion
- By evaluating the velocity function at the location of the particles.
```

peterrum commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,120 @@
+
+We note that much greater accuracy could be obtained by using a fourth
+order Runge-Kutta method or another appropriate scheme for the time integration
+of the motion of the particles. However, this does not alter which capacities
+are displayed in the present step.
```
```
does not alter which capacities are displayed
```
sounds odd...

peterrum commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,120 @@
+associated with the particles. Consequently, users can build their own
+particle solver and attribute the desired properties to the particles (e.g. mass,
+  diameter, temperature, etc.). In the present tutorial, this is used to
+  store the value of the fluid velocity and the processor id to which the particles
```
```suggestion
  store the value of the fluid velocity and the process id to which the particles
```

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+#include <deal.II/particles/generators.h>
+
+// Since the particles do not form a triangulation, they have their
+// own specific data out class which will enable us to write them
```
```suggestion
// own specific DataOut class which will enable us to write them
```

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+// Trilinos and PETSC distributed vectors
+#define FORCE_USE_OF_TRILINOS
+
+namespace LA
+{
+#if defined(DEAL_II_WITH_PETSC) && !defined(DEAL_II_PETSC_WITH_COMPLEX) && \
+  !(defined(DEAL_II_WITH_TRILINOS) && defined(FORCE_USE_OF_TRILINOS))
+  using namespace dealii::LinearAlgebraPETSc;
+#  define USE_PETSC_LA
+#elif defined(DEAL_II_WITH_TRILINOS)
+  using namespace dealii::LinearAlgebraTrilinos;
+#else
+#  error DEAL_II_WITH_PETSC or DEAL_II_WITH_TRILINOS required
+#endif
+} // namespace LA
```
Would it be possible to rather use our own vector class `LinearAlgebra::distributed::Vector`?

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+    // This class consists largely of member variables that
+    // describe the details of the particle tracking simulation and its
+    // discretization. The following parameters are about where output should
+    // land, the spatial discretization of the velocity (the default is $Q_1$),
```
land?

gassmoeller replied:
I think this is not yet addressed, maybe `where output should be written` or `stored`?

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+
+  // There remains the task of declaring what run-time parameters we can accept
+  // in input files. Since we have a very limited number of parameters, all
+  // parameters are declared in the same category.
```
```suggestion
  // parameters are declared in the same section.
```

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+
+
+    // Rene you would be more proefficient than me to write this
+    unsigned int cell_weight(
```
Are you sure this has to be public?

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+  // @sect4{Constructor}
+
+  // Constructors and destructors are rather trivial. They are very similar
+  // to what is done in step-40. we set the set of processors we want to work on
```
```suggestion
  // to what is done in step-40. We set the set of processors we want to work on
```

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+      &                                                                  cell,
+    const typename parallel::distributed::Triangulation<dim>::CellStatus status)
+  {
+    if (cell->is_active() && !cell->is_locally_owned())
+      return 0;
```
Could you explain the logic?

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+    // there)
+    // 2. How to pack the particles before shipping data around
+    // 3. How to unpack the particles after repartitioning
+    // Attach the correct functions to the signals inside
```
New line?

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+    particles_dof_handler.distribute_dofs(particles_fe);
+
+    // We generate the necessary bounding boxes for the particles generator
+    // These bounding boxes are required to quickly identify in which
+    // processors and which cell the inserted particle lies.
```
... which process and which cell owns...

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+  // using an analytically defined velocity field. This is a relatively trivial
+  // usage of the particles.
+  template <int dim>
+  void ParticleTracking<dim>::euler_analytical(double dt)
```
```suggestion
  void ParticleTracking<dim>::euler_step_analytical(double dt)
```

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+  // We integrate the particle trajectories by interpolating the value of the
+  // velocity field at the degrees of freedom to the position of the particles.
+  template <int dim>
+  void ParticleTracking<dim>::euler_interpolated(double dt)
```
```suggestion
  void ParticleTracking<dim>::euler_step_interpolated(double dt)
```

peterrum commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,745 @@
+        const auto pic = particle_handler.particles_in_cell(cell);
+        for (; particle != pic.end(); ++particle)
+          {
+            const auto &reference_location = particle->get_reference_location();
```
Is the reference location automatically updated?

peterrum commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,114 @@
+line, the program will try to read the file "`parameters.prm`" by default, and
+will execute the two dimensional version of the code.
+
+Regardless of the specific parameter file name, if the specified file does not
+exist, when you execute the program you will get an exception that no such file
+can be found:
+
+@code
+----------------------------------------------------
+Exception on processing:
+
+--------------------------------------------------------
+An error occurred in line <74> of file <../source/base/parameter_acceptor.cc> in function
+    static void dealii::ParameterAcceptor::initialize(const std::string &, const std::string &, const ParameterHandler::OutputStyle, dealii::ParameterHandler &)
+The violated condition was:
+    false
+Additional information:
+    You specified <parameters.prm> as input parameter file, but it does not exist. We created it for you.
+--------------------------------------------------------
+
+Aborting!
+----------------------------------------------------
+@endcode
+
+However, as the error message already states, the code that triggers the
+exception will also generate the specified file ("`parameters.prm`" in this case).
```
Not needed!

peterrum commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,114 @@
+We notice that, by default, the simulation runs the particle tracking with
+an analytical velocity for 2000 iterations, then runs the particle tracking with
+velocity interpolationn for the same duration. The results are written every
+10 iterations.
```
```suggestion
10th iteration.
```

peterrum commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,114 @@
+
+The directory in which this program is run contains a example parameter file by defualt.
+If you do not specify a parameter file as an argument on the command
+line, the program will try to read the file "`parameters.prm`" by default, and
```
```suggestion
line, the program will try to read the file "parameters.prm" by default, and
```

peterrum commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,114 @@
+<h1>Results</h1>
+
+The directory in which this program is run contains a example parameter file by defualt.
```
```suggestion
The directory in which this program is run contains an example parameter file by default.
```

peterrum commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,114 @@
+this case the particles would need to have additional properties such as their mass,
+and their diameter.
+- Coupling to a flow solver. This step could be straightforwardly coupled to any parallel
+steps in which the Stokes or the Navier-Stokes equations are solved (e.g. step-57)
```
You could add a reference to step-70.

gassmoeller replied:
and step-32

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,114 @@
+
+<h3>Possibilities for extensions</h3>
+
+This steps highlights some of the main capabilities of particle, notably their
```
```suggestion
This steps highlights some of the main capabilities of particles, notably their
```

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,114 @@
+capacity to be use in distributed parallel simulations. However, this step could
+be exteded in numerous manners:
+- High-order time integration (for example using a Runge-Kutta 4 method) could be
+used to increase the accuracy and allow for an increased time-steps.
```
```suggestion
used to increase the accuracy and allow for an increased time-step size.
```

gassmoeller commented on examples/step-68/doc/tooltip:
```diff
@@ -0,0 +1 @@
+A patricle tracking problem using an analytically defined velociyt field
```
```suggestion
A particle tracking problem using an analytically defined velocity field
```

blaisb commented:
Please do not review this commit. I might (possibly) have identified a bug in insert_global_particles and that hindered my progression with adding the properties to the particles.

blaisb commented:
This branch is still not ready for final review. I am sorry. I am facing what appears to be a bug with the data out of particles in parallel if the particles have properties, I will open an issue

blaisb commented:
Dear @peterrum  and @gassmoeller,
I think this is a first finished version of step-68 with all the features I wanted in it. Major changes : 
- Removed all dependency on trilinos and petsc. Now everything is done using deal.II distributed vectors
- Added properties to the particles which are also postprocessed.

I would greatly welcome your comments and your reviews? I don't plan on adding new features, so I think this is ready for a final round of review. All of your comments are more than welcomed :)

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,119 @@
+where $T$ is the period of the flow. The velocity profile in 2D ($\textbf{u}=[u,v]^T$) is :
+@f{eqnarray*}
+   u &=&  - \frac{\partial\Psi}{\partial y} = \frac{1}{\pi} \sin^2 (\pi x) \sin (\pi y) \cos (\pi y)  \cos \left( \pi \frac{t}{T} \right)\\
+   v &=&  \frac{\partial\Psi}{\partial x} = \frac{1}{\pi} \cos(\pi x) \sin(\pi x) \sin^2 (\pi y) \cos \left( \pi \frac{t}{T} \right)
```
Its hard to read the formulas in text form, but is there a factor of 2pi missing in both u and v (from the outer and inner derivative of the product rule of `sin^2(pi y)` and `sin^2(pi x)` respectively)? Or is it missing in the stream function if the velocity should remain unchanged?

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,119 @@
+  Adaptive Mesh Refinement for Geodynamic Computations." Geochemistry, Geophysics,
+   Geosystems 19.9 : 3596-3604.
+
+ <li>Blais, Bruno, et al. (2013) "Dealing with more than two materials in the FVCF–ENIP method."
```
In the test you cite Blais 2014, is that the same paper as this one? Maybe the year got mixed up?

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,90 @@
+<h1>Results</h1>
+
+The directory in which this program is run contains an example parameter file by default.
```
This is not true at the moment, please add the example prm file.

bangerth replied:
Is this still an issue? If not, please "resolve"

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,90 @@
+The directory in which this program is run contains an example parameter file by default.
+If you do not specify a parameter file as an argument on the command
+line, the program will try to read the file "parameters.prm" by default, and
+will execute the two dimensional version of the code.
```
I dont think there is any other version than 2D is there?

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,90 @@
+
+We note that, by default, the simulation runs the particle tracking with
+an analytical velocity for 2000 iterations, then runs the particle tracking with
+velocity interpolationn for the same duration. The results are written every
```
```suggestion
velocity interpolation for the same duration. The results are written every
```

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,90 @@
+
+<h3> Dynamic load balancing </h3>
+
+The following animation shows the impact of dynamic load balancing. We clearly
```
This section comes as somewhat of a surprise, because load balancing is not discussed in the introduction or elsewhere outside of the code. Do you want me to write a paragraph on why it is important when handling particles in distributed triangulations?

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,90 @@
+
+<h3>Possibilities for extensions</h3>
+
+This steps highlights some of the main capabilities of particles, notably their
```
```suggestion
This steps highlights some of the main capabilities for handling particles in deal.II, notably their
```

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,90 @@
+this case the particles would need to have additional properties such as their mass,
+and their diameter.
+- Coupling to a flow solver. This step could be straightforwardly coupled to any parallel
+steps in which the Stokes (step-32, step-70) or the Navier-Stokes equations are solved (e.g. step-57)
```
```suggestion
steps in which the Stokes (step-32, step-70) or the Navier-Stokes equations are solved (e.g. step-57).
```

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,90 @@
+<h3>Possibilities for extensions</h3>
+
+This steps highlights some of the main capabilities of particles, notably their
+capacity to be use in distributed parallel simulations. However, this step could
```
```suggestion
capacity to be used in distributed parallel simulations. However, this step could
```

blaisb commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,119 @@
+and Peter Munch (Technical University of Munich and Helmholtz-Zentrum Geesthacht)
+</i>
+
+@dealiiTutorialDOI{10.5281/zenodo.3829064,https://zenodo.org/badge/DOI/10.5281/zenodo.3829064.svg}
```
You are absolutely right.
I felt that the parallel load balancing, although very interesting, appeared out of the blue. I really love your three suggestions as they are very didactic and that would improve the quality of the step for users who are not experience in particle tracking. I will give a shot at parallel particle generation and you can give a shot at load balancing and particle transfer like you suggested. I will also integrate your other comments :). Thank you very much for taking the time to review and contribute this much :)!

blaisb commented:
@gassmoeller 
Thank you so much for your comments. I added all of your suggestions and modified the notation. I also added a new section in the introduction called :  **Challenges related to distributed particle simulations**

In this section, I added the discussion on the global generation of the particles. The two elements missing would be load balancing and particle exchange between the processors. I think that adding this to the tutorial makes the step a lot easier to follow and a lot more didactic. I will be waiting for your two paragraph and I think that after this the step should be finished and ready for final review :). I am very happy about the result.

blaisb commented:
ping @gassmoeller

gassmoeller commented:
Sorry for the radio silence. I sent my paragraphs in https://github.com/lethe-cfd/dealii/pull/3 a while ago, but I did not mention that here so it got lost. I will read through the tutorials as it is right now, my changes in the other PR are pretty self-contained.

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,149 @@
+an array which can be used to store any arbitrary number of properties
+associated with the particles. Consequently, users can build their own
+particle solver and attribute the desired properties to the particles (e.g. mass, charge
+  diameter, temperature, etc.). In the present tutorial, this is used to
+  store the value of the fluid velocity and the process id to which the particles
+  belong.
```
```suggestion
diameter, temperature, etc.). In the present tutorial, this is used to
store the value of the fluid velocity and the process id to which the particles
belong.
```

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,149 @@
+
+<h3>Challenges related to distributed particle simulations</h3>
+
+Although the present step is not computationnaly intensive, simulations that
```
```suggestion
Although the present step is not computationally intensive, simulations that
```

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,149 @@
+<h3>Challenges related to distributed particle simulations</h3>
+
+Although the present step is not computationnaly intensive, simulations that
+include particles can be computationnaly demanding and require parallelization.
```
```suggestion
include particles can be computationally demanding and require parallelization.
```

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,149 @@
+Generating the particles is not straightforward since the processor to which they belong
+must first be identified before the cell in which they are located is found.
+Deal.II provides numerous capabilities to generate particles through the Particles::Generator namespace.
+Some of these particle generator generate particles on the locally own subdomain. For example,
```
```suggestion
Some of these particle generators generate particles on the locally owned subdomain. For example,
```

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,149 @@
+does not constitute a large portion of the computational cost. For these occasions, deal.II provides
+convenient Particles::Generators that can globally insert the particles even if they are not located
+on the subdomain from which they are created. The generators first locate on which subdomain the particles
+are situated, identify is which cell they are located and exchange the necessary information amongst the processors
```
```suggestion
are situated, identify in which cell they are located and exchange the necessary information amongst the processors
```

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,149 @@
+are situated, identify is which cell they are located and exchange the necessary information amongst the processors
+to ensure that the particle is generated with the right properties. Consequently, this type of particle generation
+is communication intensive. The Particles::Generators::dof_support_points and the Particles::Generators::quadrature_points
+generate particles using a, possible non-matching, triangulation and the points of an associated dof_handler or quadrature respectively.
```
```suggestion
generate particles using a, possibly non-matching, triangulation and the points of an associated dof_handler or quadrature respectively.
```

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,149 @@
+generate particles using a, possible non-matching, triangulation and the points of an associated dof_handler or quadrature respectively.
+Furthermore, the Particles::ParticleHandler class provides the Particles::ParticleHandler::insert_global_particles function
+which enables the global insertion of particles from a vector of points and a global vector of bounding boxes. In the present step,
+we use the Particles::Generators::quadrature_points to insert the particle in the shape of a circle.
```
```suggestion
we use the Particles::Generators::quadrature_points to insert the particle in the shape of a circle, 
which is generated from an otherwise unused spherical triangulation.
```

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,149 @@
+the dynamics of a particular vortical flow : the Rayleigh-Kotte Vortex. This flow pattern
+is generally used as a complex test case for interface tracking methods
+(e.g. volume-of-fluid and level set approches) since
+it leads to strong rotation and elongation of the fluid (Blais 2013).
```
```suggestion
it leads to strong rotation and elongation of the fluid (Blais, 2013).
```

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,86 @@
+line, the program will try to read the file "parameters.prm" by default, and
+will execute the code.
+
+On any number of core, the simulation output will look like:
```
```suggestion
On any number of cores, the simulation output will look like:
```

gassmoeller commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,86 @@
+@endcode
+
+We note that, by default, the simulation runs the particle tracking with
+an analytical velocity for 2000 iterations, then runs the particle tracking with
```
```suggestion
an analytical velocity for 2000 iterations, then restarts from the beginning and runs the particle tracking with
```

gassmoeller commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,149 @@
+   v &=&  \frac{\partial\Psi}{\partial x} = 2 \cos(\pi x) \sin(\pi x) \sin^2 (\pi y) \cos \left( \pi \frac{t}{T} \right)
+@f}
+
+
```
Maybe add a sentence like the following explaining that the flow will reverse periodically and material will end up where it started:

```
It can be seen that this velocity reverses periodically due to the term $\cos \left( \pi \frac{t}{T} \right)$ and that material will end up at its starting position after every period of length $t=2T$. We will run this tutorial program for exactly one period and compare the final particle location to the initial location to illustrate this flow property.

blaisb commented:
> Sorry for the radio silence. I sent my paragraphs in [lethe-cfd#3](https://github.com/lethe-cfd/dealii/pull/3) a while ago, but I did not mention that here so it got lost. I will read through the tutorials as it is right now, my changes in the other PR are pretty self-contained.

My bad, I am sorry. I did not get that pull request warning (or it got lost through the mail).
I will work on integrating your comments right away :)!.

blaisb commented:
@gassmoeller  @peterrum 
I integrated all of your suggestions. I also rebased the code on the latest deal.II version and I did some additional cleaning (e.g. fixing spelling mistakes and syntax in some places. I think this is ready for the final review :)! 

Sorry for the delay in merging  your pull request @gassmoeller

peterrum commented:
/rebuild

gassmoeller commented:
No problem at all, I did not notice that it was not merged either :-). This now looks ready from my side.

blaisb commented:
Ping :), I'd just like this not to die :)

peterrum reviewed (approved):
@gassmoeller and I have already approved this new tutorial. I thinks it is great and ready to be merged. I think we could merge this PR next week; if there should be additional request changes, we can still do these till the next release.

drwells reviewed (changes requested):
There are still a couple of open remarks from @gassmoeller which should be addressed before we merge. I added a few more minor comments.

The largest issue with the current setup is that the style of the comments in the body of the example (i.e., those inside `step-68.cc`) don't match our usual convention of using full sentences to explain what is going on. I think this is worth fixing now and it should be pretty quick to convert the fragments into prose.

drwells commented:
The inline YouTube videos look good so I think once we address some of these style and spelling issues we can merge.

blaisb commented:
> There are still a couple of open remarks from @gassmoeller which should be addressed before we merge. I added a few more minor comments.
> 
> The largest issue with the current setup is that the style of the comments in the body of the example (i.e., those inside `step-68.cc`) don't match our usual convention of using full sentences to explain what is going on. I think this is worth fixing now and it should be pretty quick to convert the fragments into prose.

Perfect. I will take care of fixing that. It should not be too much work. If it makes the tutorial more inline with the other ones, no problem with making those minor changes

drwells commented:
Oh, one more thing I forgot to recommend - we definitely need a changelog entry so everyone gets credit for the new tutorial.

blaisb commented:
> Oh, one more thing I forgot to recommend - we definitely need a changelog entry so everyone gets credit for the new tutorial.

Sure thing. I totally forgot about that.
Is this considered a minor or a major entry? Where should I put it?
dealii/doc/news/changes/minor or
dealii/doc/news/changes/major ?
I feel that is not mine to judge :).

tjhei commented:
> Is this considered a minor or a major entry?

Use major.

blaisb commented:
> > Is this considered a minor or a major entry?
> 
> Use major.

Perfect.
Forgive my lack of knowledge, but it's my first time adding a tutorial. Which file do I need to modify to ensure that the new step-68 would appear in the build/doc/doxygen/deal.II/Tutorial.html and the graph therein? 
I'd like to prevent needless pull requests for minor stuff like that :).

masterleinad commented:
> Which file do I need to modify to ensure that the new step-68 would appear in the build/doc/doxygen/deal.II/Tutorial.html and the graph therein?

We built the documentation as part of the indentation check so you could either build it yourself or download it from https://github.com/dealii/dealii/suites/1197075587/artifacts/17677192. That being said, it already shows up on the tutorial page but is not linked to anything else. You should add something to `doc/doxygen/tutorial/tutorial.h.in` and populate `examples/step-68/doc/builds-on`.

blaisb commented:
- Added the changelog entry. I put various since the step has been the collaborative work of four authors. Toni (my PhD student) helped me start the step. Rene contributed extensively with significant major contributions. Peter had a very large role in the review of the step and his comments really made the step easier to understand.
- Added a reference to step-68 in the tutorial as suggested by @masterleinad . I used advanced techniques as the category
- Fixed spelling mistakes. I also tried to improve the sentences in the source file itself so that it is more similar to the other tutorials. I am sorry if there are mistakes remaining, as you most likely have noticed, English is not my native language and it shows in those things. Feel free to suggest any modifications, I will gladly accept them @drwells  :)!

bangerth commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,234 @@
+It can be seen that this velocity reverses periodically due to the term
+$\cos \left( \pi \frac{t}{T} \right)$ and that material will end up at its
+starting position after every period of length $t=2T$. We will run this tutorial
+program for exactly one period and compare the final particle location to the
```
I would have expected that you end back where you were at `t=T`. This would also avoid the ambiguity in the second sentence: is "one period" equal to T or 2T? In line 189 above, you seem to suggest that a period is T, but stuff only ends up where it started after 2T...

blaisb replied:
Yeah that was confusing. I kept the things the same, just mentioned that T is the half period instead. I prefer to keep it that way since this is generally how we write this flow profile in the literature.

bangerth commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,234 @@
+@f{eqnarray*}
+   u &=&  - \frac{\partial\Psi}{\partial y} = -2 \sin^2 (\pi x) \sin (\pi y) \cos (\pi y)  \cos \left( \pi \frac{t}{T} \right)\\
+   v &=&  \frac{\partial\Psi}{\partial x} = 2 \cos(\pi x) \sin(\pi x) \sin^2 (\pi y) \cos \left( \pi \frac{t}{T} \right)
+@f}
```
Do you have a picture of this flow field? If so, might be useful to show it here.

blaisb replied:
I added a youtube movie  :)!

bangerth commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,234 @@
+solution as the velocity for each particle. Therefore in this model there is no
+error in the assigned velocity to the particles, and any deviation of particle
+positions from the analytical position at a given time results from the error
+in solving the equation of motion for the particle. In the second model the
```
```suggestion
in solving the equation of motion for the particle inexactly, using a time stepping method. In the second model the
```

bangerth commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,234 @@
+in solving the equation of motion for the particle. In the second model the
+analytical velocity field is first interpolated to a finite-element vector
+space (to simulate the case that the velocity was obtained from solving a
+finite-element problem). This finite-element "solution" is then evaluated at
```
```suggestion
finite-element problem, in the same way as the ODE for each particle in step-19 depends on a finite element
solution). This finite-element "solution" is then evaluated at
```

bangerth commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,234 @@
+
+
+
+<h3>References</h3>
```
If you replace all references above by `@cite`, you don't need this section any more.

bangerth reviewed (commented):
Comments about the introduction.

My apologies for the long delay -- the start of the semester has been a bitch :-(

bangerth commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,88 @@
+Repartitioning triangulation after particle generation
+Writing particle output file: interpolated-particles-0
+Writing background field file: background-0
+Writing particle output file: ainterpolated-particles-10
```
Should this have been
```suggestion
Writing particle output file: interpolated-particles-10
```
?

bangerth commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,88 @@
+
+@htmlonly
+<p align="center">
+  <iframe width="560" height="500" src="https://youtu.be/EbgS5Ch35Xs"
```
Nice one!

bangerth commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,88 @@
+
+<h3>Possibilities for extensions</h3>
+
+This steps highlights some of the main capabilities for handling particles in deal.II, notably their
```
```suggestion
This program highlights some of the main capabilities for handling particles in deal.II, notably their
```

bangerth commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,88 @@
+used to increase the accuracy and allow for an increased time-step size.
+- The full equation of motion (with inertia) could be solved for the particles. In
+this case the particles would need to have additional properties such as their mass,
+and their diameter.
```
```suggestion
as in step-19, and if one wanted to also consider interactions with the fluid, their diameter.
```

bangerth commented on examples/step-68/doc/results.dox:
```diff
@@ -0,0 +1,88 @@
+this case the particles would need to have additional properties such as their mass,
+and their diameter.
+- Coupling to a flow solver. This step could be straightforwardly coupled to any parallel
+steps in which the Stokes (step-32, step-70) or the Navier-Stokes equations are solved (e.g. step-57).
```
```suggestion
program in which the Stokes (step-32, step-70) or the Navier-Stokes equations are solved (e.g., step-57).
```

bangerth reviewed (commented):
And some on the results section.

blaisb commented on doc/doxygen/tutorial/tutorial.h.in:
```diff
@@ -916,6 +922,12 @@
+ *   <tr valign="top">
+ *     <td> Particles
+ *     </td>
+ *     <td>step-68</td>
```
Absolutely. I'll assume from now on that step-19 is merged before step-68 :)

blaisb commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,234 @@
+ number of applications. Particles are routinely used
+ as massless tracers to visualize the dynamic of a transient flow. They
+ can also play an intrinsic role as part of a more complex finite element
+ model, as is the case for the Particle-In-Cell (PIC) method (Gassmöller et al. 2018)
```
Did not know we could do that. I like it much better this way :)!

blaisb commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,234 @@
+"load balancing", which means we estimate the computational load that is
+associated with each cell and its particles. Repartitioning the mesh then
+accounts for this combined computational load instead of the simplified
+assumption of the number of cells.
```
I think you mean the 2018 paper? I added a citation to it there.

bangerth replied:
The appropriate paper, whichever one it is :-)

blaisb commented:
I applied all of Wolfgang's comment. Thank you very much for taking the time to review the tutorial and for the suggestions :)! The tutorial now assumes that step-19 exists (it depends on step-19). I think it is ready to be merged, but that's your call. I  would merge it after step-19 though :)!
@peterrum @bangerth  

That was fun work, once this one is merged I will start work on another example.

peterrum commented:
Do you want to to squash your commits?

blaisb commented:
> Comments about the introduction.
> 
> My apologies for the long delay -- the start of the semester has been a bitch :-(

Don

> Do you want to to squash your commits?

"Gonna squash em all, gonna squash em all...." 
Sorry. I squashed all of the commits from September into one. Hopefully this should do it
:)!

blaisb commented:
> 
> > Do you want to to squash your commits?
> 
> "Gonna squash em all, gonna squash em all...."
> Sorry. I squashed all of the commits from September into one. Hopefully this should do it
> :)!

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+
+// Since the particles do not form a triangulation, they have their
+// own specific DataOut class which will enable us to write them
+// to commonly used parallel vtu format:
```
```suggestion
// to commonly used parallel vtu format (or any number of other file formats):
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+  //
+  // The ParameterAcceptor paradigm requires all parameters to be writable by
+  // the ParameterAcceptor methods. In order to avoid bugs that would be very
+  // difficult to track down (such as writing things like `time = 0` instead of
```
```suggestion
  // difficult to track down (such as writing things like `if (time = 0)` instead of
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+  // The ParameterAcceptor paradigm requires all parameters to be writable by
+  // the ParameterAcceptor methods. In order to avoid bugs that would be very
+  // difficult to track down (such as writing things like `time = 0` instead of
+  // `time == 0`), we declare all the parameters in an external class, which is
```
```suggestion
  // `if(time == 0)`), we declare all the parameters in an external class, which is
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    // This class consists largely of member variables that
+    // describe the details of the particle tracking simulation and its
+    // discretization. The following parameters are about where output should
+    // land, the spatial discretization of the velocity (the default is $Q_1$),
```
Following @peterrum and @gassmoeller 's suggestion
```suggestion
    // written to, the spatial discretization of the velocity (the default is $Q_1$),
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    : ParameterAcceptor("Particle Tracking Problem/")
+  {
+    add_parameter(
+      "Velocity degree", velocity_degree, "", this->prm, Patterns::Integer(1));
```
Why is it that this one call is different from the others in that it gets `this->prm` and `Patterns::Integer(1)` as arguments? Is the `prm` a default argument and you need to specify it because you want to provide a pattern? I think that some of the other arguments could also benefit from a pattern.

blaisb replied:
Indeed, that was a clear mistake on my part. I fixed it and added patterns for the doubles and integer as well as a min when it is pertinent.

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    Vortex()
+      : Function<dim>(dim)
+    {}
+    virtual void vector_value(const Point<dim> &point,
```
```suggestion

    virtual void vector_value(const Point<dim> &point,
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+                              Vector<double> &  values) const override;
+  };
+
+  template <int dim>
```
```suggestion

  template <int dim>
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    const double T = 4;
+    // The velocity profile for the Rayleigh-Kothe vertex is time-dependent.
+    // Consequently, the current time in the
+    // simulation (t) must be gathered from the Function object.
```
it works better (=makes it easier to read) if you move the comment above the entire function, because then the whole function shows up in the same highlighted block in the html page

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+
+  private:
+    // The particles_generation function is responsible for the initial
+    // generation of the particles on top of the background grid
```
```suggestion
    // generation of the particles on top of the background grid.
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    // When the velocity profile is interpolated to the position of the
+    // particles, it must first be stored using degrees of freedom.
+    // Consequently, as is the case for other parallel case (e.g. step-40) we
+    // initialize the degrees of freedom on the background grid
```
```suggestion
    // initialize the degrees of freedom on the background grid.
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    // initialize the degrees of freedom on the background grid
+    void setup_background_dofs();
+
+    // In one of the test case, the function is mapped to the background grid
```
```suggestion
    // In one of the test cases, the function is mapped to the background grid
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    // The next two functions are responsible for carrying out step of explicit
+    // Euler time integration for the cases where the velocity field is
+    // interpolated at the positions of the particles or calculated
+    // analytically, respectively
```
```suggestion
    // analytically, respectively.
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    // interpolated at the positions of the particles or calculated
+    // analytically, respectively
+    void euler_step_interpolated(double dt);
+    void euler_step_analytical(double dt);
```
mark the input arguments of these two functions as `const`

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    // results for the particles and for the velocity profile on the background
+    // mesh, respectively.
+    void output_particles(unsigned int it);
+    void output_background(unsigned int it);
```
same here with the input arguments

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    // how the domain needs to be partitioned so that every MPI rank receives a
+    // roughly equal amount of work (potentially not an equal number of cells).
+    // While the function is called from the outside, it is connected to the
+    // corresponding signal from inside this class, therefore it can be private.
```
```suggestion
    // corresponding signal from inside this class, therefore it can be `private`.
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    void euler_step_interpolated(double dt);
+    void euler_step_analytical(double dt);
+
+    // The cell_weight() function indicates to the triangulation how much
```
```suggestion
    // The `cell_weight()` function indicates to the triangulation how much
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    void output_background(unsigned int it);
+
+    // The private members of this class are similar to other parallel deal.II
+    // examples. The parameters are stored as a const member. It is important
```
```suggestion
    // examples. The parameters are stored as a `const` member. It is important
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+
+    // The private members of this class are similar to other parallel deal.II
+    // examples. The parameters are stored as a const member. It is important
+    // to note that we keep the Vortex class as a member since its time
```
```suggestion
    // to note that we keep the `Vortex` class as a member since its time
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+
+  // The constructors and destructors are rather trivial. They are very similar
+  // to what is done in step-40. We set the processors we want to work on
+  // to all machines available (MPI_COMM_WORLD) and
```
```suggestion
  // to all machines available (`MPI_COMM_WORLD`) and
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    , fluid_fe(FE_Q<dim>(par.velocity_degree), dim)
+    , mapping(par.velocity_degree)
+    , pcout(
+        {std::cout, Utilities::MPI::this_mpi_process(mpi_communicator) == 0})
```
```suggestion
        std::cout, Utilities::MPI::this_mpi_process(mpi_communicator) == 0)
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+  private:
+    // The particles_generation function is responsible for the initial
+    // generation of the particles on top of the background grid
+    void particles_generation();
```
That's a slightly funny name for a function. Function names should have a verb in it to indicate an *action*. May I suggest `generate_particles()`?

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+  void ParticleTracking<dim>::particles_generation()
+  {
+    // We create an hyper_cube triangulation which we globally define. This
+    // triangulation englobes the full trajectory of the particles.
```
```suggestion
    // triangulation covers the full trajectory of the particles.
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+
+    // We create a particle triangulation which is solely used to generate
+    // the points which will be used to insert the particles. This
+    // triangulation is an hyper_shell which is off-set from the
```
```suggestion
    // triangulation is a hyper shell which is offset from the
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    unsigned int n_locally_owned_cells = 0;
+    for (const auto &cell : particle_triangulation.active_cell_iterators())
+      if (cell->is_locally_owned())
+        n_locally_owned_cells++;
```
I think that this whole loop is equivalent to writing `particle_triangulation.n_locally_owned_cells()`.

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    // points quadrature. Consequently, one particle will be generated
+    // at the centroid of each cell.
+    Particles::Generators::quadrature_points(particle_triangulation,
+                                             QGauss<dim>(1),
```
This might be easier to understand:
```suggestion
                                             QMidpoint<dim>(),
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+
+  // @sect4{Background DOFs and interpolation}
+
+  // This function sets up the background degree of freedom used for the
```
```suggestion
  // This function sets up the background degrees of freedom used for the
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+  void ParticleTracking<dim>::setup_background_dofs()
+  {
+    fluid_dh.distribute_dofs(fluid_fe);
+    IndexSet locally_owned_dofs = fluid_dh.locally_owned_dofs();
```
```suggestion
    const IndexSet locally_owned_dofs = fluid_dh.locally_owned_dofs();
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    fluid_dh.distribute_dofs(fluid_fe);
+    IndexSet locally_owned_dofs = fluid_dh.locally_owned_dofs();
+    IndexSet locally_relevant_dofs;
+    DoFTools::extract_locally_relevant_dofs(fluid_dh, locally_relevant_dofs);
```
didn't we introduce a function where you can get this index set returned from the dof handler in the same way as for the locally owned dofs?

blaisb replied:
Looking at :
https://www.dealii.org/developer/doxygen/deal.II/classDoFHandler.html
I think there are no such functions. I can always be wrong though :)!.

bangerth replied:
No, you're almost certainly right. We might have *wanted* to do that, but didn't. (See #10563 for something I might have had in mind.)

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+
+  // This function takes care of interpolating the
+  // vortex velocity field to the field vector. This is achieved rather easily
+  // by using the <code>VectorTools::interpolate</code> function.
```
doxygen will automatically do the markup:
```suggestion
  // by using the VectorTools::interpolate() function.
```

bangerth commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    // Looping over all particles in the domain using a particle iterator
+    for (auto particle = particle_handler.begin();
+         particle != particle_handler.end();
+         ++particle)
```
write this as
```
  for (const auto &particle : particle_handler)
```

bangerth reviewed (commented):
First set of comments on the .cc file. Rest later today. Only small stuff, should be easy to address!

drwells commented:
I think we are at the point (like in step-69) where, nearing 200 comments, we are likely to start breaking GitHub. @bangerth if there are more large change sets could you collect them into a PR against this branch on lethe-cfd?

bangerth commented:
Yes, good point. I keep forgetting. My remaining edits are all here: https://github.com/lethe-cfd/dealii/pull/4

blaisb commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,810 @@
+    background_triangulation.signals.pre_distributed_repartition.connect(
+      std::bind(
+        &Particles::ParticleHandler<dim>::register_store_callback_function,
+        &particle_handler));
```
Sorry for my lack of knowledge, but how can I do that?

bangerth replied:
I think this could work (give it a shot):
```
background_triangulation.signals.pre_distributed_repartition.connect(
  [particle_handler]() { particle_handler.register_store_callback_function(); });
```

blaisb replied:
Sorry for pestering you with this, but I must acknowledge that my knowledge of lambda functions is very limited here. I get the following error message trying your solution : 

`home/blaisb/work/dealii/dealii/dealii/examples/step-68/step-68.cc: In member function ‘void Step68::ParticleTracking<dim>::generate_particles()’:
/home/blaisb/work/dealii/dealii/dealii/examples/step-68/step-68.cc:419:9: error: capture of non-variable ‘Step68::ParticleTracking<dim>::particle_handler’
  419 |       [&particle_handler]() {
      |         ^~~~~~~~~~~~~~~~
/home/blaisb/work/dealii/dealii/dealii/examples/step-68/step-68.cc:278:47: note: ‘dealii::Particles::ParticleHandler<dim> Step68::ParticleTracking<dim>::particle_handler’ declared here
  278 |     Particles::ParticleHandler<dim>           particle_handler;
      |                                               ^~~~~~~~~~~~~~~~
/home/blaisb/work/dealii/dealii/dealii/examples/step-68/step-68.cc: In lambda function:
/home/blaisb/work/dealii/dealii/dealii/examples/step-68/step-68.cc:420:9: error: ‘this’ was not captured for this lambda function
  420 |         particle_handler->register_store_callback_function();
      |         ^~~~~~~~~~~~~~~~
`

blaisb commented:
@bangerth  I applied all of your comments but there are two points that remain unresolved. Could you look at them in the file tab of this PR? :) Thank you so much for the review.

drwells reviewed (approved):
:+1:

masterleinad commented on examples/step-68/step-68.cc:
```diff
@@ -0,0 +1,835 @@
+        const typename parallel::distributed::Triangulation<dim>::CellStatus
+          status) -> unsigned int { return this->cell_weight(cell, status); });
+
+    background_triangulation.signals.pre_distributed_repartition.connect(
+      std::bind(
+        &Particles::ParticleHandler<dim>::register_store_callback_function,
+        &particle_handler));
```
```suggestion
    background_triangulation.signals.pre_distributed_repartition.connect(
      [this]() {this->particle_handler.register_store_callback_function(); });
```

blaisb replied:
This works for the register_store_callback_function.
However, the register_load_callback_function expects an argument in the form of a serialization (see https://www.dealii.org/current/doxygen/deal.II/classParticles_1_1ParticleHandler.html#a9a79889c1ad3b569a405488d4bab8263)
How do I deal with this? Because 
`
    background_triangulation.signals.post_distributed_repartition.connect(
      [this]() {
        this->particle_handler.register_load_callback_function();
      });
`
does not work obviously since it is missing an argument. Sorry again for asking, but my experience with signals or lambdas is extremely limited.

peterrum replied:
Cannot you simply write?
```cpp
    background_triangulation.signals.post_distributed_repartition.connect(
      [&](){this->particle_handler.register_load_callback_function(false)});
```

blaisb replied:
Ok this seems to compile well. But now the code segfaults in parallel when partitioning. I'll try to debug and see what happens. Some change somehow must be broken things.

peterrum commented:
@blaisb There seems to be a merge conflict here. Could you fix it. After that I think it will be merged!

blaisb commented:
> @blaisb There seems to be a merge conflict here. Could you fix it. After that I think it will be merged!

Yeah no problem, I will take care of that tomorrow morning first thing. I'll message you after that.

drwells commented:
@bangerth does that comment signify your approval with the current state? If so lets merge!

bangerth commented:
Yes! I've got to run, which is typically a bad sign to commit or merge anything, but happy to let someone else do it who can devote a minute or two of thought to the process!

drwells commented on examples/step-68/doc/intro.dox:
```diff
@@ -0,0 +1,232 @@
+Toni El Geitani Nehme (Polytechnique Montréal),
+Rene Gassmöller (University of California Davis),
+and Peter Munch (Technical University of Munich and Helmholtz-Zentrum Geesthacht).
+Bruno Blais was supported by NSERC Discovery grant
```
It took me a minute to realize that this is not a typo for NERSC :)

blaisb commented:
> If you can't figure it out, just leave it as you had it and one of us takes care of it as a follow-up.

I figured it out.
It now runs in debug mode in parallel at any number of cores on two of my machines.
There were some minor issues with vectors and the interpolated euler loop.

blaisb commented:
I would like to take some time to thank you all for the reviews. It was a great learning opportunity for me and I am glad I was able to make a durable contribution to a library that has really been a blessing since I started my career as a professor. Thank you for your time. I know we are all swamped by so many tasks and I appreciate every moment of the time you took to help.
@drwells @bangerth @peterrum @gassmoeller

bangerth commented:
The pleasure is all ours! We hope for more contributions from you and your group! :-)

peterrum commented:
For some reason the videos are not displayed in my case properly? Does anyone have the same issue? step-19 is fine!?

blaisb commented:
> For some reason the videos are not displayed in my case properly? Does anyone have the same issue? step-19 is fine!?

Yeah they are not displayed correctly on mine also. Is it a Youtube issue? I used the same code as in the other steps.
